beacon/beacon.go

```diff
@@ -848,21 +879,42 @@ func (pd *ProtocolDriver) runConsensusPhase(ctx context.Context, epoch types.Epo
 		undecided proposalList
 		err       error
 	)
-	for round := types.FirstRound; round < pd.config.RoundsNumber; round++ {
+
+	// First round
+	round := types.FirstRound
+	pd.setRoundInProgress(round)
+	if msg, err := pd.genFirstRoundMsgBody(epoch); err != nil {
+		logger.With().Error("failed to get firstround message", log.Err(err), round)
+	} else {
```

Comment:
The only reason `pd.genFirstRoundMsgBody` might fail is if the identity isn't active in the given epoch. But `active` already has been checked for that so can this method return an error?

Alternatively `pd.genFirstRoundMsgBody` could just be part of `sendFirstRoundVote` to simplify the logic.

Reply:
> this work is common for all signers so it makes no sense to repeat it

I initially assumed the method just creates a message and returns -> that doesn't need to be its own method. The problem here is that the method has an additional side effect (sorting proposals) that isn't not obvious from its name or on a first glance when looking at the method.

> it mutates shared state by sorting incoming proposals

That's fair, but then sorting should be done once in a method that's named "sortIncommingProposals" or similar and the instantiation of the Message object can be done in `sendFirstRoundVote`. At least I think that would make it clearer what's happening 🙂

Reply:
I removed the method entirely. Btw, I discovered that it had a bug - it took a shared Rlock and mutated the data :(